Dip/Dip/Dip.swift

```diff
@@ -42,8 +42,7 @@ public class DependencyContainer {
     case Int(IntegerLiteralType)
   }
   
-  private var dependencies = [DefinitionKey : Definition]()
-  private var lock: OSSpinLock = OS_SPINLOCK_INIT
```

Comment:
As for me using any kind of lock is bad practice and leads to more hard to catch bugs... 
Pretty silly example but still possible: someone call resolve not from main thread and then for any reason switch back to another thread in factory block and call `resolve` on that thread. It will still create a deadlock as I understand. So it looks like we exchange one set of "you do not do that this way" to another.

But if you still would like to use locks we can give it a try. Though I think there is no need to lock `register` method cause they are supposed to be called at one point in a program. The only place worth locking is probably access to resolved instances pool. What do you think?